### PR TITLE
Commerce 1955 Add methods to update order and order item pricing

### PR DESCRIPTION
**Pull request by alecsloan**

### CODE REVIEW

**Comment by alecsloan** [comment]
ci:test

**Comment by alecsloan** [comment]
@marco-leo Draft pull, as I haven't tested yet. Just for you to review.

File: commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemServiceImpl.java [hunk 1]
```diff
@@ -288,6 +293,35 @@ public CommerceOrderItem updateCommerceOrderItemInfo(
...
+			BigDecimal discountPercentageLevel2,
+			BigDecimal discountPercentageLevel3,
+			BigDecimal discountPercentageLevel4)
+		throws PortalException {
+
+		CommerceOrderItem commerceOrderItem =
+			commerceOrderItemLocalService.getCommerceOrderItem(
+				commerceOrderItemId);
+
+		CommerceOrder commerceOrder = commerceOrderItem.getCommerceOrder();
```
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemServiceImpl.java]: @alecsloan We should just check if the user can update the order using the order permission, that already checks for guest visibility

**Review by marco-leo — CHANGES_REQUESTED** [review]

File: commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemServiceImpl.java (hunk 1 above)
Comment (alecsloan) [reply commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemServiceImpl.java]: In the upsert method commercePrdocutViewPermission still has to be checked to see if they can view the product?

**Comment by alecsloan** [comment]
ci:test

File: commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderServiceImpl.java [hunk 2]
```diff
@@ -493,6 +496,48 @@ public CommerceOrder updateCommerceOrder(
...
+			BigDecimal totalDiscountPercentageLevel1,
+			BigDecimal totalDiscountPercentageLevel2,
+			BigDecimal totalDiscountPercentageLevel3,
+			BigDecimal totalDiscountPercentageLevel4)
+		throws PortalException {
+
+		CommerceOrder commerceOrder =
+			commerceOrderLocalService.getCommerceOrder(commerceOrderId);
+
+		CommerceAccountServiceUtil.getCommerceAccount(
```
Comment (marco-leo) [inline commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderServiceImpl.java]: @alecsloan why we are not using _commerceOrderModelResourcePermission.check(
			getPermissionChecker(), commerceOrderItem.getCommerceOrderId(),
			ActionKeys.UPDATE);
 ?

**Comment by alecsloan** [comment]
ci:test

File: commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java [hunk 3]
```diff
@@ -557,6 +557,32 @@ public CommerceOrderItem updateCommerceOrderItemPrice(
...
+			BigDecimal finalPrice, BigDecimal discountPercentageLevel1,
+			BigDecimal discountPercentageLevel2,
+			BigDecimal discountPercentageLevel3,
+			BigDecimal discountPercentageLevel4)
+		throws PortalException {
+
+		CommerceOrderItem commerceOrderItem =
+			commerceOrderItemPersistence.findByPrimaryKey(commerceOrderItemId);
+
+		commerceOrderItem.setUnitPrice(unitPrice);
```
Comment (ethanbustad) [inline commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java:574]: Hey @alecsloan I think we should also be updating the `modifiedDate` here.

File: commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderLocalServiceImpl.java [hunk 4]
```diff
@@ -985,6 +985,66 @@ public CommerceOrder updateCommerceOrder(
...
+			BigDecimal totalDiscountPercentageLevel1,
+			BigDecimal totalDiscountPercentageLevel2,
+			BigDecimal totalDiscountPercentageLevel3,
+			BigDecimal totalDiscountPercentageLevel4)
+		throws PortalException {
+
+		CommerceOrder commerceOrder = commerceOrderPersistence.findByPrimaryKey(
+			commerceOrderId);
+
+		commerceOrder.setLastPriceUpdateDate(new Date());
```
Comment (ethanbustad) [inline commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderLocalServiceImpl.java:1012]: also `modifiedDate` here.

File: commerce-service/src/main/resources/resource-actions/default.xml [hunk 5]
```diff
@@ -8,6 +8,7 @@
 			<supports>
 				<action-key>MANAGE_COMMERCE_AVAILABILITY_ESTIMATES</action-key>
 				<action-key>MANAGE_COMMERCE_COUNTRIES</action-key>
+				<action-key>MANAGE_COMMERCE_ORDER_PRICES</action-key>
```
Comment (ethanbustad) [inline commerce-service/src/main/resources/resource-actions/default.xml:11]: We should add a new lang key for this.

**Review by ethanbustad — CHANGES_REQUESTED** [review]

File: commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java (hunk 3 above)
Comment (alecsloan) [reply commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java:574]: That's not auto handled by persistence?

**Comment by alecsloan** [comment]
ci:test

File: commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java (hunk 3 above)
Comment (alecsloan) [reply commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java:574]: Looks like it should be:
```
if (!commerceOrderModelImpl.hasSetModifiedDate()) {
	if (serviceContext == null) {
		commerceOrder.setModifiedDate(now);
	}
	else {
		commerceOrder.setModifiedDate(
			serviceContext.getModifiedDate(now));
	}
}

Comment (ethanbustad) [reply commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java:574]: Woah no way! That didn't use to be there :)

**Pull request merged**